services/hub.go

```diff
@@ -0,0 +1,463 @@
+			}
+		case <-h.close:
+			return
+		}
+	}
+}
+
+func (h *Hub) StartEventWatcher() {
+	fn := h.watchEventDeliveriesCollection()
+	err := h.watchCollectionFn(fn, mongo.Pipeline{}, m.EventDeliveryCollection, h.close)
+	if err != nil {
+		log.WithError(err).Error("database collection watcher exited unexpectedly")
+	}
+}
+
+type CLIEvent struct {
+	UID              string                `json:"uid"`
+	ForwardedHeaders httpheader.HTTPHeader `json:"forwarded_headers" bson:"forwarded_headers"`
+	Data             json.RawMessage       `json:"data"`
+
+	// for filtering this event delivery
+	EventType string `json:"-"`
+	DeviceID  string `json:"-"`
+	AppID     string `json:"-"`
+	GroupID   string `json:"-"`
+}
+
+func (h *Hub) watchEventDeliveriesCollection() func(doc map[string]interface{}) error {
+	return func(doc map[string]interface{}) error {
+		var ed *datastore.EventDelivery
+		b, err := json.Marshal(doc)
+		if err != nil {
+			return err
+		}
+
+		err = json.Unmarshal(b, &ed)
+		if err != nil {
+			return err
+		}
+
+		if ed.CLIMetadata == nil {
+			return nil
+		}
+
+		h.events <- &CLIEvent{
+			UID:              ed.UID,
+			Data:             ed.Metadata.Data,
+			ForwardedHeaders: ed.Headers,
+			EventType:        ed.CLIMetadata.EventType,
+			AppID:            ed.AppID,
+			DeviceID:         ed.DeviceID,
+			GroupID:          ed.GroupID,
+		}
+
+		return nil
+	}
+}
+
+func (h *Hub) StartRegister() {
+	for {
+		select {
+		case client := <-h.register:
+			h.lock.Lock()
+			h.deviceClients[client.deviceID] = client
+			h.lock.Unlock()
+		case <-h.close:
+			return
+		}
+	}
+}
+
+func (h *Hub) StartUnregister() {
+	for {
+		select {
+		case client := <-h.unregister:
+			h.lock.Lock()
+			delete(h.deviceClients, client.deviceID)
+			h.lock.Unlock()
+		case <-h.close:
+			return
+		}
+	}
+}
+
+func (h *Hub) Stop() {
+	close(h.close)
+}
+
+func (h *Hub) UpdateEventDeliveryStatus(id string) {
+	ed, err := h.eventDeliveryRepo.FindEventDeliveryByID(context.Background(), id)
+	if err != nil {
+		log.WithError(err).WithField("event_delivery_id", id).Error("failed to find event delivery")
+	}
+
+	err = h.eventDeliveryRepo.UpdateStatusOfEventDelivery(context.Background(), *ed, datastore.SuccessEventStatus)
+	if err != nil {
+		log.WithError(err).WithField("event_delivery_id", id).Error("failed to update event delivery status")
+	}
+}
+
+type ListenRequest struct {
+	HostName   string   `json:"host_name"`
+	DeviceID   string   `json:"device_id"`
+	SourceID   string   `json:"source_id"`
+	EventTypes []string `json:"event_types"`
+}
+
+type LoginRequest struct {
+	HostName string `json:"host_name"`
+	DeviceID string `json:"device_id"`
+}
+
+type LoginResponse struct {
+	Device *datastore.Device      `json:"device"`
+	Group  *datastore.Group       `json:"group"`
+	App    *datastore.Application `json:"app"`
+}
+
+type AckEventDelivery struct {
+	UID string `json:"uid"`
+}
+
+func (h *Hub) LoginHandler(w http.ResponseWriter, r *http.Request) {
```

Comment:
We can potentially make this better by:
1. Decoupling the websocket server http handlers away from the `hub`.
2. Using contexts to pass the `hub` down to the handlers where necessary.